docs/design/2020-01-24-collations.md

```diff
@@ -0,0 +1,243 @@
+| t     | CREATE TABLE `t` (                                          |
+|       |   `a` varchar(20) COLLATE utf8mb4_general_ci NOT NULL,      |
+|       |   PRIMARY KEY (`a`)                                         |
+|       | ) ENGINE=InnoDB DEFAULT CHARSET=utf8mb4 COLLATE=utf8mb4_bin |
++-------+-------------------------------------------------------------+
+1 row in set
+tidb> insert into t values ('A');
+Query OK, 1 row affected
+tidb> insert into t values ('a');
+Query OK, 1 row affected // Should report error "Duplicate entry 'a'"
+```
+In the case above, the user creates a column with a case-insensitive collation `utf8mb4_general_ci`. Since the 'real' collation of TiDB is always `utf8mb4_bin`, inserting data into the primary key with "A" and "a" does not lead to an error.
+
+What's more, the collation `utf8mb4_bin` in MySQL is defined with attribute `PAD SPACE`, that is, trailing spaces are ignored when comparison. Currently `utf8mb4_bin` in TiDB is with attribute `NO PAD`:
+
+```
+tidb> create table t1(a varchar(20) charset utf8mb4 collate utf8mb4_bin key);
+Query OK, 0 rows affected
+tidb> insert into t1 values ('a')
+Query OK, 1 row affected
+tidb> insert into t1 values ('a ');
+Query OK, 1 row affected //Should report error "Duplicate entry 'a '"
+```
+
+#### Requirements
+
+In this proposal, both of the compatibility issues can be solved in the new version of TiDB, and with the help of the version of metadata in TiDB, we can always indicate if a collation is old(the "fake one") or new(the "real one"). However, here we propose several requirements below and careful considerations are needed to meet them as many as possible:
+  1. For an existing TiDB cluster, its behavior remains unchanged after upgrading to newer version
+    - a. After upgrade, the behavior of existing tables remains unchanged.
+    - b. After upgrade, the behavior of newly-created tables remains unchanged.
+    - c. If a TiDB cluster is replicating the data to a MySQL instance through Binlog, all  replications remain unchanged.
+  2. Users of MySQL who requires collations can move to TiDB without modifying their scripts/applications.
+  3. Users of TiDB can distinguish the tables with old collation and new collations easily, if they both exist in a TiDB cluster.
+  4. All replications/backups/recovers through TiDB tools like Binlog-related mechanism, CDC and BR should not encounter unexpected errors caused from the changes of collations in TiDB. For example, 
+    - When using BR, if backup data from old cluster is applied to the new cluster, the new cluster should know that the data is binary collated without padding, or otherwise 'Duplicate entry' error may be reported for primary key/unique key columns.
+    - If we don't allow both old collations and new collations exist in a TiDB cluster(the Option 4, see below), trying to make replications/backup & recovery between old/new clusters will break this prerequisite.
```

Comment:
I think there should be a requirement that you can fix bugs in a collation. Because indexed values affect sort order, MySQL is not able to fix a collation **ever**. They instead need to create a new collation, and migration is painful.  You should expect that there are bugs in the implementation, and have a plan to be able to scan through and upgrade incorrect values.

Reply:
Thank you, I think this can be resolved by marking an internal version in TiDB's collation metadata.

I will try to add some descriptions about this.

Reply:
> I think this can be resolved by marking an internal version in TiDB's collation metadata.
> 

The real difficulty here is how to cope with `index`, should we need to regenerate index?

Reply:
Of course, index reorgnization seems to be the only way to fix it.

Reply:
If we can fix bug by this, why we still introduce concept `0900`? We can almost use the lastest UCA(now is 10.0), and take the upgradation as a big bug fix. For example, we can use `utf8mb4_ai_ci` and don't expose the version  to user.

Reply:
> For example, we can use utf8mb4_ai_ci and don't expose the version to user.

The key part of this is the upgrading: when we upgrade the internal version of `utf8mb4_ai_ci`, the existing on-disk data must be updated too. I've updated the RFC on the implementation of this, PTAL

Reply:
@bb7133  Then what‘s the problem with `utf8mb4_ai_ci` without the version?

Reply:
I just mean that we may have bugs in the implementation and write the wrong data on disk.